ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryExecutor.java

```diff
@@ -647,4 +671,107 @@ private static void updateListProperty(
     valueList.add(value);
     properties.put(key, valueList);
   }
+
+  private static class TransientQuerySinkProcessor implements
+      Processor<Object, GenericRow, Void, Void> {
+
+    private final TransientQueryQueue queue;
+    private final Optional<ImmutableMap<TopicPartition, Long>> endOffsets;
+    private final ConcurrentHashSet<TopicPartition> donePartitions;
+    private ProcessorContext<Void, Void> context;
+
+    TransientQuerySinkProcessor(
+        final TransientQueryQueue queue,
+        final Optional<ImmutableMap<TopicPartition, Long>> endOffsets,
+        final ConcurrentHashSet<TopicPartition> donePartitions) {
+
+      this.queue = queue;
+      this.endOffsets = endOffsets;
+      this.donePartitions = donePartitions;
+    }
+
+    @Override
+    public void init(final ProcessorContext<Void, Void> context) {
+      Processor.super.init(context);
+      this.context = context;
+      if (endOffsets.isPresent()) {
+        // registering a punctuation here is the easiest way to check
+        // if we're complete even if we don't see all records.
+        // The in-line check in process() is faster,
+        // but some records may be filtered out before reaching us.
+        context.schedule(
+            Duration.ofMillis(100),
+            PunctuationType.WALL_CLOCK_TIME,
+            timestamp -> checkCompletion()
+        );
+      }
+    }
+
+    @Override
+    public void process(final Record<Object, GenericRow> record) {
+      final Optional<TopicPartition> topicPartition =
+          context
+              .recordMetadata()
+              .map(m -> new TopicPartition(m.topic(), m.partition()));
+
+      final boolean alreadyDone =
+          topicPartition.isPresent() && donePartitions.contains(topicPartition.get());
+
+      // We ignore null values in streams as invalid data and drop the record.
+      // If we're already done with the current partition, we just drop the record.
+      if (record.value() != null && !alreadyDone) {
+        queue.acceptRow(null, record.value());
+      }
+
+      checkCompletion();
```

Comment:
For every record added to the queue, we check the endOffsets across all partitions in `checkCompletion`. I am wondering if we can make it cheaper (might not be needed if this is already very cheap) by just checking for completion in the partition we are currently reading from, where the current record is read from

Reply:
That's a good observation, but it should actually be pretty cheap already, since this is only checking local variables. I'd be surprised if just checking the current record's partition is noticeably faster, but I'll go ahead and do it.